SpeedyTransforms/src/matrix_transform.jl

```diff
@@ -0,0 +1,128 @@
+    S = SpectralTransform(spectrum, grid; NF, ArrayType, nlayers, LegendreShortcut, architecture)
+
+    npoints = get_npoints(grid)
+    nharmonics = LowerTriangularArrays.nonzeros(spectrum)
+    forward = on_architecture(architecture, zeros(Complex{NF}, nharmonics, npoints))
+    backward = on_architecture(architecture, zeros(Complex{NF}, npoints, nharmonics))
+
+    field = on_architecture(architecture, zeros(NF, grid))
+    coeffs = on_architecture(architecture, zeros(Complex{NF}, spectrum))
+
+    forward_matrix!(forward, S, field, coeffs)
+    backward_matrix!(backward, S, field, coeffs)
+
+    # SCRATCH MEMORY FOR FOURIER NOT YET LEGENDRE TRANSFORMED AND VICE VERSA
+    scratch_memory = on_architecture(architecture, zeros(Complex{NF}, grid, nlayers).data)
+
+    # PRECOMPUTE GRADIENT AND INTEGRATION MATRICES
+    gradients = gradient_arrays(NF, spectrum)
+
+    return MatrixSpectralTransform{
+        NF,
+        typeof(architecture),
+        ArrayType_,
+        typeof(spectrum),
+        typeof(grid),
+        typeof(coslat),
+        typeof(forward),
+        typeof(gradients),
+    }(
+        architecture,
+        spectrum, grid, nlayers,
+        coslat, coslat⁻¹,
+        S.norm_sphere,
+        forward,
+        backward,
+        scratch_memory,
+        gradients,
+    )
+end
+
+# TODO Algorithm to compute the actual transform matrices
+forward_matrix!(F, S::AbstractSpectralTransform, field::AbstractField, coeffs::LowerTriangularArray) = nothing
+backward_matrix!(B, S::AbstractSpectralTransform, field::AbstractField, coeffs::LowerTriangularArray) = nothing
+
+function transform!(                        # GRID TO SPECTRAL
+        coeffs::LowerTriangularArray,       # output: spectral coefficients
+        field::AbstractField,               # input: gridded values
+        scratch_memory,                     # explicit scratch memory (not used only in spectral to grid)
+        M::MatrixSpectralTransform,         # precomputed spectral transform
+    )
+    LinearAlgebra.mul!(coeffs.data, M.forward, field.data)
+    return coeffs
+end
+
+function transform!(                        # SPECTRAL TO GRID
+        field::AbstractField,               # gridded output
+        coeffs::LowerTriangularArray,       # spectral coefficients input
+        scratch_memory,                     # explicit scratch memory to use
+        M::MatrixSpectralTransform;         # precomputed transform
+        unscale_coslat::Bool = false,       # unscale with cos(lat) on the fly?
+    )
+    # use first n layers of scratch memory
+    scratch = view(scratch_memory, :, 1:size(coeffs, 2))
+    
+    # multiply into scratch which is complex typed and then take real part into field
+    # imaginary part should be zero but destination is used to store intermediate results
+    # explicitly convert to real also for NaN + NaN*im results
+    LinearAlgebra.mul!(scratch, M.backward, coeffs.data)
+    field.data .= real.(scratch)
```

Comment:
I am always a bit confused about broadcasting operations like that. Is this allocating because of executing `real.(scratch)` first or not?